### PR TITLE
Add PUT /manager/configuration and /cluster/{node_id}/configuration endpoints

### LINKED ISSUE #7181: Add /manager/configuration and /cluster/{node_id}/configuration endpoints
|Wazuh version|Component|
|---|---|
| 4.1.0 | Wazuh API |

Hello team,

as part of the epic issue https://github.com/wazuh/wazuh/issues/7148, we need to split the `manager/files` and `/cluster/{node_id}/files` endpoints.

This issue will cover configuration files. For that, we need the following new endpoints:
- [x] `PUT /manager/configuration`
  - XML format required.
  - **RBAC:** Action `manager:update_config`. Resourceless.
- [x] `PUT /cluster/{node_id}/configuration`
  - XML format required.
  - **RBAC:** Action `cluster:update_config`. Resource `node:id`.

These new endpoints require adaptations in the framework in order to make a proper content validation and only perform over the `ossec.conf` file.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

Regards,
Víctor Fernández.

### PR DESCRIPTION
**Pull request by CarlosRS9**
Hello team,

This PR closes #7181. This PR adds the following:

- Two new endpoints to allow the user to update the wazuh configuration (ossec.conf) using the Wazuh API:
  - `PUT /manager/configuration`
  - `PUT /cluster/{node_id]/configuration`.
- Unit and integration tests for the new endpoints.
- A fix for a typo in the Spec.yaml

### Unit tests

wazuh/core/tests/
```
================================= test session starts =================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 643 items                                                                   

framework/wazuh/core/tests/test_active_response.py ............                 [  1%]
framework/wazuh/core/tests/test_agent.py ...................................... [  7%]
............................................................................... [ 20%]
..................................                                              [ 25%]
framework/wazuh/core/tests/test_cdb_list.py .................................   [ 30%]
framework/wazuh/core/tests/test_common.py .......                               [ 31%]
framework/wazuh/core/tests/test_configuration.py .............................. [ 36%]
......                                                                          [ 37%]
framework/wazuh/core/tests/test_decoder.py ................                     [ 39%]
framework/wazuh/core/tests/test_input_validator.py ...                          [ 40%]
framework/wazuh/core/tests/test_logtest.py ..                                   [ 40%]
framework/wazuh/core/tests/test_manager.py ................................     [ 45%]
framework/wazuh/core/tests/test_ossec_queue.py ...................              [ 48%]
framework/wazuh/core/tests/test_pyDaemonModule.py ......                        [ 49%]
framework/wazuh/core/tests/test_results.py .................................... [ 54%]
....                                                                            [ 55%]
framework/wazuh/core/tests/test_rootcheck.py ..........                         [ 57%]
framework/wazuh/core/tests/test_rule.py ......................                  [ 60%]
framework/wazuh/core/tests/test_syscollector.py ...                             [ 60%]
framework/wazuh/core/tests/test_utils.py ...................................... [ 66%]
............................................................................... [ 79%]
............................................................................... [ 91%]
..............                                                                  [ 93%]
framework/wazuh/core/tests/test_wazuh_socket.py ....................            [ 96%]
framework/wazuh/core/tests/test_wdb.py .....................                    [100%]

===================== 643 passed, 11 warnings in 2.84 seconds =====================
```
wazuh/core/cluster/tests/
```
================================= test session starts =================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 55 items                                                                    

framework/wazuh/core/cluster/tests/test_cluster.py ...................          [ 34%]
framework/wazuh/core/cluster/tests/test_common.py .......                       [ 47%]
framework/wazuh/core/cluster/tests/test_control.py sssss                        [ 56%]
framework/wazuh/core/cluster/tests/test_local_client.py .                       [ 58%]
framework/wazuh/core/cluster/tests/test_utils.py .......                        [ 70%]
framework/wazuh/core/cluster/tests/test_worker.py ............ss.s              [100%]

================== 47 passed, 8 skipped, 9 warnings in 0.48 seconds ===================
```

core/cluster/dapi/tests/
```
================================= test session starts =================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 22 items                                                                    

framework/wazuh/core/cluster/dapi/tests/test_dapi.py ......................     [100%]

======================= 22 passed, 11 warnings in 0.62 seconds ========================
```

wazuh/tests/
```
================================= test session starts =================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 672 items                                                                   

framework/wazuh/tests/test_active_response.py .........                         [  1%]
framework/wazuh/tests/test_agent.py ........................................... [  7%]
.........................................                                       [ 13%]
framework/wazuh/tests/test_cdb_list.py ........................................ [ 19%]
.......                                                                         [ 20%]
framework/wazuh/tests/test_ciscat.py ..                                         [ 21%]
framework/wazuh/tests/test_cluster.py .....ss                                   [ 22%]
framework/wazuh/tests/test_decoders.py ...............................          [ 26%]
framework/wazuh/tests/test_group.py ........                                    [ 27%]
framework/wazuh/tests/test_logtest.py ......                                    [ 28%]
framework/wazuh/tests/test_manager.py ......................................... [ 34%]
                                                                                [ 34%]
framework/wazuh/tests/test_mitre.py ........................................... [ 41%]
............................................................................... [ 53%]
..........................................................                      [ 61%]
framework/wazuh/tests/test_rootcheck.py ....................................... [ 67%]
..........                                                                      [ 69%]
framework/wazuh/tests/test_rule.py ............................................ [ 75%]
...........                                                                     [ 77%]
framework/wazuh/tests/test_sca.py .......                                       [ 78%]
framework/wazuh/tests/test_security.py ........................................ [ 84%]
.............................                                                   [ 88%]
framework/wazuh/tests/test_stats.py .............                               [ 90%]
framework/wazuh/tests/test_syscheck.py ........................                 [ 94%]
framework/wazuh/tests/test_syscollector.py ............                         [ 95%]
framework/wazuh/tests/test_task.py ............................                 [100%]

=============== 670 passed, 2 skipped, 15 warnings in 72.56 seconds ===============
```

wazuh/rbac/tests/
```
================================= test session starts =================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 224 items                                                                   

framework/wazuh/rbac/tests/test_auth_context.py ..                              [  0%]
framework/wazuh/rbac/tests/test_decorators.py ................................. [ 15%]
........................................................................        [ 47%]
framework/wazuh/rbac/tests/test_default_configuration.py ...................... [ 57%]
...............................                                                 [ 71%]
framework/wazuh/rbac/tests/test_orm.py ........................................ [ 89%]
.............                                                                   [ 95%]
framework/wazuh/rbac/tests/test_preprocessor.py ...........                     [100%]

====================== 224 passed, 1 warnings in 230.25 seconds =======================
```

### Integration tests
```
test_cluster_endpoints.tavern.yaml 
         53 passed, 1 xpassed, 1 warnings

test_manager_endpoints.tavern.yaml 
         34 passed, 1 warnings

test_rbac_black_cluster_endpoints.tavern.yaml 
         20 passed, 1 warnings

test_rbac_black_manager_endpoints.tavern.yaml 
         18 passed, 1 warnings

test_rbac_white_cluster_endpoints.tavern.yaml 
         20 passed, 1 warnings

test_rbac_white_manager_endpoints.tavern.yaml 
         18 passed, 1 warnings

test_security_DELETE_endpoints.tavern.yaml 
         15 passed, 1 warnings

test_security_GET_endpoints.tavern.yaml 
         18 passed, 1 warnings

test_security_POST_endpoints.tavern.yaml 
         7 passed, 1 warnings

test_security_PUT_endpoints.tavern.yaml 
         8 passed, 1 warnings

```

### CODE REVIEW

File: framework/wazuh/core/configuration.py [hunk 1]
```diff
@@ -805,3 +806,19 @@ def get_active_configuration(agent_id, component, configuration):
     else:
         raise WazuhError(1117 if "No such file or directory" in rec_msg or "Cannot send request" in rec_msg else 1116,
                          extra_message='{0}:{1}'.format(component, configuration))
+
+
+def write_ossec_conf(new_conf: str):
+    """
+    Replaces the current wazuh configuration (ossec.conf) with the provided configuration.
```
Comment (davidjiglesias) [inline framework/wazuh/core/configuration.py:813]:
```suggestion
    Replace the current wazuh configuration (ossec.conf) with the provided configuration.
```

File: framework/wazuh/manager.py [hunk 2]
```diff
@@ -384,3 +384,53 @@ def get_basic_info():
...
+    result = AffectedItemsWazuhResult(all_msg=f"Configuration was successfully updated"
+                                              f"{' in specified node' if node_id != 'manager' else ''}",
+                                      some_msg='Could not update configuration in some nodes',
+                                      none_msg=f"Could not update configuration"
+                                               f"{' in specified node' if node_id != 'manager' else ''}"
+                                      )
+    backup_conf = None
+    try:
+        # Check a configuration has been provided
+        if new_conf is None or not new_conf or len(new_conf) == 0:
```
Comment (davidjiglesias) [inline framework/wazuh/manager.py:409]:
```suggestion
        if not new_conf:
```

File: framework/wazuh/manager.py [hunk 3]
```diff
@@ -384,3 +384,53 @@ def get_basic_info():
...
+            raise WazuhError(1125)
+
+        # Check if the configuration is valid
+        new_conf = prettify_xml(new_conf)
+
+        # Create a backup of the current configuration before attempting to replace it
+        backup_conf = load_wazuh_xml(common.ossec_conf)
+        write_ossec_conf(new_conf)
+        is_valid = validate_ossec_conf()
```
Comment (davidjiglesias) [inline framework/wazuh/manager.py:418]: Can't we validate before writing?

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: framework/wazuh/manager.py (hunk 3 above)
Comment (CarlosRS9) [reply framework/wazuh/manager.py:418]: No, unfortunately due to the way the Core validation process works we cannot send the configuration itself to the socket. We need to place the new configuration inside the ossec.conf and then request its validation using the socket instead.

Comment (CarlosRS9) [reply framework/wazuh/manager.py:418]: In the future we should improve this to be able to send the configuration to be validated directly to the socket itself, avoiding unnecesary changes to the ossec.conf and potential issues derived from that.

File: api/api/spec/spec.yaml [hunk 4]
```diff
@@ -200,6 +200,14 @@ x-rbac-catalog:
         actions: ['cluster:read']
         resources: ['node:id:worker1', 'node:id:worker3']
         effect: "deny"
+    'cluster:update_config':
+      description: "Update current Wazuh's node configuration"
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply api/api/spec/spec.yaml]: Changed

File: api/test/integration/env/configurations/rbac/manager/black_config.txt [hunk 5]
```diff
@@ -6,4 +6,5 @@
 | manager:read_file        | file:path |        | *                            |
 | manager:upload_file      | *         | *      |                              |
 | manager:restart          | *         | *      |                              |
+| manager:upload_config    | *         |        | *                            |
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply api/test/integration/env/configurations/rbac/manager/black_config.txt]: done

File: api/test/integration/env/configurations/rbac/cluster/black_config.txt [hunk 6]
```diff
@@ -13,4 +13,5 @@
 | cluster:read_file        | node:id&file:path |                     | master-node,worker2 & etc/ossec.conf,etc/rules/local_rules.xml,ruleset/rules/0350-amazon_rules.xml             |
 | cluster:upload_file      | node:id           |                     | master-node,worker2                                                                                            |
 | cluster:delete_file      | node:id&file:path |                     | master-node,worker2 & etc/decoders/0005-wazuh_decoders.xml, etc/ossec.conf                                     |
+| cluster:upload_config    | node:id           | worker1             | master-node,worker2                                                                                            |
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply api/test/integration/env/configurations/rbac/cluster/black_config.txt]: done

File: framework/wazuh/manager.py (hunk 3 above)
Comment (davidjiglesias) [reply framework/wazuh/manager.py:418]: Apply file backup system

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: framework/wazuh/manager.py (hunk 3 above)
Comment (CarlosRS9) [reply framework/wazuh/manager.py:421]: Added file backup functionality

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Apply back_up changes

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**